docker.go

```diff
@@ -1377,3 +1344,156 @@ func (p *DockerProvider) getDefaultNetwork(ctx context.Context, cli *client.Clie
 
 	return reaperNetwork, nil
 }
+
+func (p *DockerProvider) isConnectedToRemoteDaemon() (bool, error) {
+	daemonUrl, err := url.Parse(p.client.DaemonHost())
+	if err != nil {
+		return false, err
+	}
+
+	switch daemonUrl.Scheme {
+	case "unix", "npipe":
+		return false, nil
+	case "http", "https", "tcp":
+		return true, err
+	default:
+		return false, nil
+	}
+}
+
+// mapToDockerMounts maps the given []ContainerMount to the corresponding
+// []mount.Mount for further processing
+func (p *DockerProvider) mapToDockerMounts(ctx context.Context, containerMounts ContainerMounts) ([]mount.Mount, error) {
+	mounts := make([]mount.Mount, 0, len(containerMounts))
+
+	remoteDaemon, err := p.isConnectedToRemoteDaemon()
+	if err != nil {
+		return nil, err
+	}
+
+	if p.runningInContainer {
+		if err := p.initContainerEnvInformation(ctx); err != nil {
+			return nil, err
+		}
+	}
+
+	for idx := range containerMounts {
+		m := containerMounts[idx]
+
+		if bindMount, ok := m.Source.(GenericBindMountSource); ok && p.runningInContainer && !remoteDaemon {
+			var remapped bool
+		BindMountMapping:
+			for i := range p.containerEnv.BindMounts {
+				hostBindMount := p.containerEnv.BindMounts[i]
+
+				// check if part of the bind source is part of a host-mount into this container
+				if strings.Contains(bindMount.HostPath, hostBindMount.Target.Target()) {
+					bindMount.HostPath = strings.Replace(bindMount.HostPath, hostBindMount.Target.Target(), hostBindMount.Source.Source(), 1)
+					m.Source = bindMount
+					remapped = true
+					break BindMountMapping
+				}
+			}
+			if !remapped {
+				return nil, fmt.Errorf("cannot bind mount %s in nested container because it's not mounted from host", bindMount.HostPath)
+			}
+		}
+
+		var mountType mount.Type
+		if mt, ok := mountTypeMapping[m.Source.Type()]; ok {
+			mountType = mt
+		} else {
+			continue
+		}
+
+		containerMount := mount.Mount{
+			Type:     mountType,
+			Source:   m.Source.Source(),
+			ReadOnly: m.ReadOnly,
+			Target:   m.Target.Target(),
+		}
+
+		switch typedMounter := m.Source.(type) {
+		case BindMounter:
+			containerMount.BindOptions = typedMounter.GetBindOptions()
+		case VolumeMounter:
+			containerMount.VolumeOptions = typedMounter.GetVolumeOptions()
+		case TmpfsMounter:
+			containerMount.TmpfsOptions = typedMounter.GetTmpfsOptions()
+		}
+
+		mounts = append(mounts, containerMount)
+	}
+
+	return mounts, nil
+}
+
+func (p *DockerProvider) initContainerEnvInformation(ctx context.Context) error {
+	var initErr error
+
+	p.initContainerEnvOnce.Do(func() {
+		remote, err := p.isConnectedToRemoteDaemon()
+		if err != nil {
+			initErr = err
+			return
```

Comment:
:smile: yeah that is not really a beauty but it's a limitation of `sync.Once`.
The `initErr` is declared outside of the `sync.Once` closure and is returned at the very end.
Every error occurring within the `sync.Once` closure is assigned to the `initErr` to allow the outer function returning the error in case there was one.
The `return` in the next line does not return the error but only returns from the closure.

You know what I mean?

Reply:
Ah I got it, Probably because I was using GH for the review I did not understand each scope properly. Thanks for the clarification!

Reply:
```golang
p.initContainerEnvOnce.Do(func() {
```

It's a function 🤦  Thanks again for your patience

Reply:
Yeah, the API of `sync.Once` is not that great :smile: 
I would love to have something like `sync.Once{}.DoSuccessfully(func() error{ ... }))` to be able to repeat it if it erros or so but yeah....maybe in the future :smile: but I totally get the confusion especially when using GitHub for reviewing